prov/verbs/src/windows/verbs_nd_ibv.c

```diff
@@ -0,0 +1,857 @@
+/*
+ * Copyright (c) 2022 Canon Medical Research USA, Inc.  All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <assert.h>
+#include "ndspi.h"
+#include "fi_verbs_nd.h"
+
+struct ibv_device **ibv_get_device_list(int *num_devices)
+{
+	trace(__func__);
+	if (num_devices) {
+		assert(vrb_nd_infra.adapters.count <= INT_MAX);
+		*num_devices = (int)vrb_nd_infra.adapters.count;
+	}
+	return vrb_nd_infra.adapters.device_list;
+}
+
+void ibv_free_device_list(struct ibv_device **list)
+{
+	trace(__func__);
+	// Since we are keeping state statically, there is nothing to free.
+}
+
+struct ibv_context *ibv_open_device(struct ibv_device *device)
+{
+	trace(__func__);
+	return get_context_by_device(device);
+}
+
+int ibv_close_device(struct ibv_context *context)
+{
+	trace(__func__);
+	// Since we are keeping state statically, there is nothing to close.
+	return FI_SUCCESS;
+}
+
+// The IND2Adapter::Query method doesn't provide data for a number of fields
+// in the ibv_device_attr structure. The following definitions were inferred
+// from what ibv_query_device returned on a Linux implementation.
+#define VRB_ND_MAX_CQ_CNT (1 << 24)
+#define VRB_ND_MAX_QP_CNT (1 << 18)
+#define VRB_ND_MAX_SRQ_CNT (1 << 23)
+#define VRB_ND_MAX_MR_CNT (1 << 24)
```

Comment:
Fair point about the magic numbers being specific to my test hardware. Short of collecting numbers from a bunch of different hardware and using the lowest common denominator, I'm not sure what else we could do. And I don't have access to a variety of hardware.

I can be more specific about the hardware I collected the numbers from if you think that adds any value.

ND does support SRQ, so I think a non-zero value makes sense.

Reply:
If ND supports SRQ, then we probably need to figure out if the HW supports SRQ somehow.  Or does ND require SRQ support?  Because it's optional for RDMA devices.

Reply:
Btw, I don't think anything in libfabric or any users of libfabric will end up caring about most of these values.  So anything reasonable should be okay.  The max_srq is an exception that I'm aware of.